### PR TITLE
Adding Doxygen CI workflow to update documentation site

### PR DESCRIPTION
**Pull request by capnkenny**
This should allow us to automatically update the Doxygen documentation based on whatever is currently live in the master branch.

- Only will update as a PR is merged into master
- Will only push the updated documentation into the `gh-pages` branch as `Github Actions`

### CODE REVIEW

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:9]: iirc jobs have a name too, so you could just name this `Build and publish documentation`

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:19]: IMHO we should make a dedicated parameter for cibuild.sh/build.sh which *only* builds the documentation.

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:21]: If we do the above, we shouldn't need this step, which makes debugging CI faster and easier

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:31]: I believe we can set `DOXYGEN_OUTPUT_DIRECTORY` as a CMake variable to `artifacts/docs` or some such (as part of cibuild, most likely) to avoid this step entirely.

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:35]: We probably don't actually need the date here since the `git commit` step includes the date in the git log anyway. What would be more useful, IMO, is the commit hash that triggered the build, which you can get using `$GITHUB_SHA`

**Review by FiniteReality — CHANGES_REQUESTED** [review]
Looks good on paper, but I think we can do better for debugging purposes.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [inline .github/workflows/doxygen.yml:20]: I added a flag to only generate documentation... this should work with and without `cibuild.sh` as the change is rooted in `build.sh`.

Should we add this for the Windows variant as well?

File: .github/workflows/doxygen.yml
Comment (capnkenny) [inline .github/workflows/doxygen.yml:28]: Updated to use the SHA of the current commit on master (that triggers this workflow) if I've understood the docs correctly.

File: doxygen/CMakeLists.txt
Comment (capnkenny) [inline doxygen/CMakeLists.txt:20]: Funnily enough, we use `DOXYGEN_OUTPUT_DIR` as our own variable, but we never used to set `DOXYGEN_OUTPUT_DIRECTORY` which is what CMake uses to define where everything should go. Was this intended?

(Logic behind this is that this now allows controlling where the files go - for us, I can specify the `/docs` folder so once pushed, the gh-pages branch should be happy with the docs folder)

File: scripts/build.sh
Comment (capnkenny) [inline scripts/build.sh:223]: If the documentation flag is passed, it assumes that Generate was never called to generate the Doxyfile so we call it manually here.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml]: This _should_ be resolved.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml]: Yup! I wrote another comment wondering about that CMake var. Resolved.

File: .github/workflows/doxygen.yml
Comment (FiniteReality) [inline .github/workflows/doxygen.yml:6]: Should be `main`

File: doxygen/CMakeLists.txt
Comment (FiniteReality) [reply doxygen/CMakeLists.txt:20]: It probably should have been `DOXYGEN_OUTPUT_DIRECTORY` in the first place - I'm personally fine with doing a rename instead of anything strange like this

**Review by FiniteReality — CHANGES_REQUESTED** [review]

File: .github/workflows/doxygen.yml
Comment (tannergooding) [reply .github/workflows/doxygen.yml:6]: This should hopefully get resolved simply with rebasing or merging the PR with latest

Comment (tannergooding) [reply .github/workflows/doxygen.yml:6]: oh wait, nevermind, new file 😄

File: .github/workflows/doxygen.yml
Comment (tannergooding) [inline .github/workflows/doxygen.yml:29]: I feel like this should be creating a PR to the pages repo and applying auto-merge once CI/sign-off completes

File: .github/workflows/doxygen.yml
Comment (tannergooding) [inline .github/workflows/doxygen.yml:25]: Is there docs on this e-mail or is there some "official-ness" to it?

File: scripts/build.sh
Comment (tannergooding) [inline scripts/build.sh:31]: What's AHHHH here?

File: scripts/build.sh
Comment (tannergooding) [reply scripts/build.sh:225]: Why not just `generate=true` if `documentation` is `true`, just like we do for `CI` so it flows naturally and works with the rest of the script?

File: .github/workflows/doxygen.yml
Comment (tannergooding) [reply .github/workflows/doxygen.yml:20]: I think yes. Ideally we can do any build step on either system.

File: scripts/build.sh
Comment (capnkenny) [reply scripts/build.sh:31]: Ah, I don't remember if this was even me or not, but this was left as part of a reminder for the xports below it. Previously, we would use the default compiler, but since CI would default to GCC 9 at the time and `clang` defaulted to clang-9, we added the two export lines to ensure that we always use Clang 10 instead.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml:25]: For this, I followed the example provided from the Checkout v2 Actions Repo (from GH) - https://github.com/actions/checkout

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml:29]: Would you be able to assist with setting up auto-merge? This is something I'm very unfamiliar with 😅

Also, so I understand the flow desired:
Checkout to new feature branch
Add the docs folder that was generated
Commit to the feature branch
Push (to remote)
Request PR between the feature branch and the `gh-pages` branch

Does this seem correct?

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml:6]: Yep, gotta change this to main now

File: scripts/build.sh
Comment (capnkenny) [reply scripts/build.sh:225]: I think it's really just an oversight I had from the first implementation - the thought was to not build since it was unnecessary, we only needed to generate and then build the `Doxygen` target as needed.

I'll correct this

File: doxygen/CMakeLists.txt
Comment (capnkenny) [reply doxygen/CMakeLists.txt]: Moved all to `DOXYGEN_OUTPUT_DIRECTORY` - just kept the step that reports if it is already defined (it will print it so we know where the output is)

File: scripts/build.sh
Comment (capnkenny) [reply scripts/build.sh]: Rewrote this to follow the general flow for CI

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml]: Changed this to main

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml]: Tried to implement this (GH doesn't provide a stock "Create Pull Request" action so I had to use one from the marketplace instead.)

File: scripts/build.sh
Comment (capnkenny) [reply scripts/build.sh]: This is removed.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml]: This has been replaced - an e-mail was found as per https://github.community/t/github-actions-bot-email-address/17204/3 that will actually utilize the github-actions bot account for commits, PRs, etc.

File: .github/workflows/doxygen.yml
Comment (capnkenny) [inline .github/workflows/doxygen.yml]: In reference to this convo: https://github.com/novelrt/NovelRT/pull/257/files/046abc10878bc4185964794d304d061ccd4bda8a#r610989328

File: .github/workflows/doxygen.yml
Comment (tannergooding) [inline .github/workflows/doxygen.yml:38]: nit: If we can trivially acquire https://github.com/cli/cli, that would be better than depending on a third party CI leg
-- It may even be default on the GH Actions machines

That should just be downloading the zip from the release URL and then running `gh auth logic` (which should pick up `with: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}`)
and then doing `gh pr create --title "title" --body "body" --base gh-pages`

I imagine the `create-pull-request@1.0.0` action is doing basically that itself anyways

Comment (tannergooding) [reply .github/workflows/doxygen.yml:38]: https://github.blog/2021-03-11-scripting-with-github-cli/#using-gh-in-github-actions

It does indeed come built in

**Review by FiniteReality — CHANGES_REQUESTED** [review]
LGTM after you make the changes Tanner mentioned

File: .github/workflows/doxygen.yml
Comment (capnkenny) [reply .github/workflows/doxygen.yml:42]: Finally got around to playing with GH CLI and testing this. Latest commit should include the change iirc

**Comment by capnkenny** [comment]
Hey @tannergooding - if you could re-review when you get a chance? Thanks!

**Review by FiniteReality — APPROVED** [review]
LGTM

File: doxygen/CMakeLists.txt
Comment (FiniteReality) [inline doxygen/CMakeLists.txt:26]: Before the merge though, could somebody quickly explain why we need this again? I'm skimming through the logic and I'm not 100% sure what it's doing

Comment (capnkenny) [reply doxygen/CMakeLists.txt:26]: Hey @FiniteReality - this is setting the main page to be the contents of `README.md`, and links together the Code of Conduct and the License to the doxygen output (this was added prior to me, but removing it removed all traces of the License, and the main page became blank).

Comment (FiniteReality) [reply doxygen/CMakeLists.txt:26]: Weird. I'll have to check the git blame, I guess :sweat_smile:

**Pull request merged**
